mirrord/layer/src/tcp_steal.rs

```diff
@@ -117,19 +155,133 @@ impl TcpHandler for TcpStealHandler {
 
 impl TcpStealHandler {
     pub async fn next(&mut self) -> Option<ClientMessage> {
-        let (connection_id, value) = self.read_streams.next().await?;
-        match value {
-            Some(Ok(bytes)) => Some(ClientMessage::TcpSteal(LayerTcpSteal::Data(TcpData {
-                connection_id,
-                bytes: bytes.to_vec(),
-            }))),
-            Some(Err(err)) => {
-                error!("connection id {connection_id:?} read error: {err:?}");
-                None
+        select! {
+            opt = self.read_streams.next() => {
+                let (connection_id, value) = opt?;
+                match value {
+                    Some(Ok(bytes)) => Some(ClientMessage::TcpSteal(LayerTcpSteal::Data(TcpData {
+                        connection_id,
+                        bytes: bytes.to_vec(),
+                    }))),
+                    Some(Err(err)) => {
+                        error!("connection id {connection_id:?} read error: {err:?}");
+                        None
+                    }
+                    None => Some(ClientMessage::TcpSteal(
+                        LayerTcpSteal::ConnectionUnsubscribe(connection_id),
+                    )),
+                }
+            }
+            Some(res) = self.http_response_receiver.recv() => {
+                Some(ClientMessage::TcpSteal(LayerTcpSteal::HttpResponse(res)))
             }
-            None => Some(ClientMessage::TcpSteal(
-                LayerTcpSteal::ConnectionUnsubscribe(connection_id),
-            )),
         }
     }
+
+    /// Send a filtered HTTP request to the application in the appropriate port.
+    /// If this is the first filtered HTTP from its connection to arrive at this layer, a new
+    /// connection will be started for it, otherwise it will be sent in the existing connection.
+    pub(crate) async fn forward_request(&mut self, request: HttpRequest) -> Result<(), LayerError> {
+        if let Some(sender) = self.http_request_senders.get(&request.connection_id) {
+            sender
+                .send(request)
+                .await
+                .map_err::<HttpForwarderError, _>(From::from)?
+        } else {
+            self.create_http_connection(request).await?
+        }
+        Ok(())
+    }
+
+    /// Manage a single tcp connection, forward requests, wait for responses, send responses back.
+    async fn connection_task(
+        mut request_receiver: Receiver<HttpRequest>,
+        mut sender: SendRequest<Full<Bytes>>,
```

Comment:
Needs a more informative name

Reply:
Changed to `http_request_sender`, @eyalb181.